### PR TITLE
#12015 - Do not display bulk actions if the necessary rigths are missing

### LINKED ISSUE #12015: Do not display bulk actions associated with rights that the users do not have
### Problem Description

Currently, the expected behavior in regards to the display of the bulk actions associated with rights that the logged in users do not have is met only for the 'Messages' directory - bulk action 'Delete', 'Upload quarantine documents' in each directory, and the 'Delete/restore', 'Archive/De-archive' in the 'Events' directory and 'Event participants' tabs where if the respective rights are missing the associated actions are not displayed in the 'Bulk actions' dropdowns.

There are 2 perspectives that need to be addressed:
1. where the actions associated with the missing rights are shown in the dropdown and the actions are successful:
**- Cases directory - edit bulk actions:** 
   If the user has UserRight.CASE_EDIT the bulk actions dropdown will contain the "Edit" functionality but the actions inside the bulk edit popup should be limited based on the specific case edit rights ( same logic as there is is place when editing a case in single mode). When the user does not have a specific edit right, the checkbox should be read only.
The following rights are necessary for the specific edit actions from "Edit cases" popup window:

- [x] "Change Disease" -> CASE_CHANGE_DISEASE
- [x] "Change case classification" -> CASE_CLASSIFY
- [x] "Change investigation status" -> CASE_INVESTIGATE
- [x] "Change case outcome" -> CASE_CLASSIFY
- [x] "Change facility" -> CASE_TRANSFER

**Note:**  There is no need to set the following checkboxes to readOnly based on rights:
 - "Change case surveillance officer"  (will depend on CASE_EDIT right) 
 - "Change share with reporting tool"
 
image

2. where the actions associated with the missing rights are shown and an error message stops the user from completing the action:
- Cases directory - bulk actions: edit (change surveillance officer or all edit actions if the full set of edit existing cases rights are missing), link event (link event requires EVENT_CREATE and EVENTPARTICIPANT_CREATE rights + EVENT_SURVEILLANCE feature configuration);
- Contacts directory -link to event (requires EVENT_CREATE and EVENTPARTICIPANT_CREATE rights + EVENT_SURVEILLANCE feature configuration);
- Event participants tab - bulk actions: create contacts (requires CONTACT_CREATE right);
- Samples directory - bulk actions: delete/restore (requires SAMPLE_DELETE right);

### Proposed Change

The bulk actions displayed for the user should all be consistently displayed based on the rights of the user logged in, as we currently do for part of the delete/restore, archive/de-archive, and upload documents actions that are not displayed in some bulk actions dropdowns when the respective rights are missing.

### Added Value/Benefit

Consistency in regards to the display based on the user's rights and the display of less error messages caused by displaying actions associated with missing rights.

### Acceptance Criteria

- [x] in the dropdown of 'Bulk actions' the actions displayed should be based on the rights of the user - i.e. if a certain right is missing the action associated with it should not be displayed;
- [x] for the edit actions that contain multiple 'sub-actions' - if only some if those sub-actions are missing, the 'Edit' action should still be displayed and contain only those sub-actions that the user has rights for ( when the user does not have the proper right, the checkboxes should be readonly in the "Edit cases" popup)

Directories to be covered:
- [x] cases
- [x] contacts
- [x] event participants
- [x] samples

### Implementation Details

_No response_

### Mockups

_No response_

### Additional Information

An extensive explanation on which actions are dependent on which rights was written and can be provided to help with the development of the issue.

The display of the bulk edit mode across directories where it is implemented will be addressed by #11881

### PR DESCRIPTION
**Pull request by carina29**
Fixes #12015

### CODE REVIEW

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13018

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13018

File: sormas-ui/src/main/java/de/symeda/sormas/ui/caze/CasesView.java [hunk 1]
```diff
@@ -812,7 +812,8 @@ public HorizontalLayout createStatusFilterBar() {
 									}));
 						}
 
-						if (FacadeProvider.getFeatureConfigurationFacade().isFeatureEnabled(FeatureType.EVENT_SURVEILLANCE)) {
+						if (FacadeProvider.getFeatureConfigurationFacade().isFeatureEnabled(FeatureType.EVENT_SURVEILLANCE)
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/caze/CasesView.java:815]: in the `UiUtil` class there are som function to make these kind of checks shorter and cleaner, please use those ones and create new ones if needed

**Review by leventegal-she — CHANGES_REQUESTED** [review]

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13018

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13018

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13018

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13018

File: sormas-ui/src/main/java/de/symeda/sormas/ui/UiUtil.java [hunk 2]
```diff
@@ -19,14 +21,26 @@ public static boolean permitted(Set<FeatureType> features, UserRight userRight)
 		return enabled(features) && permitted(userRight);
 	}
 
+	public static boolean permitted(Set<FeatureType> features, Set<UserRight> userRights) {
```
Comment (leventegal-she) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/UiUtil.java:24]: this method looks to be always used with a single feature type, so why require a set of feature types here?
also why not use vararg for userRights, would be easier to use and wouldn't need to transform the set into an array in the `permitted` method

**Review by leventegal-she — CHANGES_REQUESTED** [review]

File: sormas-ui/src/main/java/de/symeda/sormas/ui/UiUtil.java (hunk 2 above)
Comment (carina29) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/UiUtil.java:24]: Yes, I've assumed that after replacing all the occurences of FacadeProvider.getFeatureConfigurationFacade().isFeatureEnabled(featureType) in the project there will be places where multiple featureTypes will be necessary, but I've checked now the code and it looks like right now always just one featureType is used, as you've said. 

I also changed the userRights to use varargs.

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13018

**Review by leventegal-she — APPROVED** [review]

**Pull request merged**